renderdoc/driver/gl/wrappers/gl_buffer_funcs.cpp

```diff
@@ -712,6 +715,11 @@ void WrappedOpenGL::glNamedBufferDataEXT(GLuint buffer, GLsizeiptr size, const v
     }
     else
     {
+      if(IsBackgroundCapturing(m_State))
```

Comment:
There's an ordering problem here but it shows up in many of the other cases - in the case where this resource has been postponed and we're doing a last-minute initial contents prepare before the write, we're actually too late. The call to `glNamedBufferDataEXT` has happened above on [line 614](https://github.com/baldurk/renderdoc/pull/1434/files#diff-04b05176fced25f4b25a0cc6ef9d5857R614). That means the buffer data has already been modified and we've missed our chance to obtain the data.

This will get a bit messy since we normally call the GL function before doing any serialised preprocessing, and this is important for e.g. making sure the serialiser has the write timestamp & duration for the call. I think you'll need to add a new bit of code above the GL function call to do the `MarkResourceFrameReferenced` at least on any functions that can possibly have a dirty-ref. That's especially annoying in the case of e.g. `Common_glTextureSubImage2DEXT` which isn't even called until after the real GL function call has been made.

Reply:
Hm.. yeah, I'll take a look into this. But yes, if the original GL call already happens, our marking as referenced in background should happen before it.